src/Planner/Planner.cpp

```diff
@@ -363,488 +1118,295 @@ void Planner::buildQueryPlanIfNeeded()
-
-                        const auto * interpolate_expression = interpolate_expression_nodes[0];
-                        if (!interpolate_expression->result_type->equals(*expression_to_interpolate->result_type))
-                        {
-                            auto cast_type_name = expression_to_interpolate->result_type->getName();
-                            Field cast_type_constant_value(cast_type_name);
-
-                            ColumnWithTypeAndName column;
-                            column.name = calculateConstantActionNodeName(cast_type_name);
-                            column.column = DataTypeString().createColumnConst(0, cast_type_constant_value);
-                            column.type = std::make_shared<DataTypeString>();
+                addTotalsHavingStep(query_plan, expression_analysis_result, query_analysis_result, planner_context, query_node, result_actions_to_execute);
+                having_executed = true;
+            }
 
-                            const auto * cast_type_constant_node = &interpolate_actions_dag->addColumn(std::move(column));
+            addCubeOrRollupStepIfNeeded(query_plan, aggregation_analysis_result, query_analysis_result, planner_context, select_query_info, query_node);
 
-                            FunctionCastBase::Diagnostic diagnostic = {interpolate_expression->result_name, interpolate_expression->result_name};
-                            FunctionOverloadResolverPtr func_builder_cast
-                                = CastInternalOverloadResolver<CastType::nonAccurate>::createImpl(std::move(diagnostic));
+            if (!having_executed && expression_analysis_result.hasHaving())
+                addFilterStep(query_plan, expression_analysis_result.getHaving(), "HAVING", result_actions_to_execute);
+        }
 
-                            ActionsDAG::NodeRawConstPtrs children = {interpolate_expression, cast_type_constant_node};
-                            interpolate_expression = &interpolate_actions_dag->addFunction(func_builder_cast, std::move(children), interpolate_expression->result_name);
-                        }
+        if (query_processing_info.isFromAggregationState())
+        {
+            if (expression_analysis_result.hasWindow())
+                throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                    "Window functions does not support processing from WithMergeableStateAfterAggregation");
+        }
+        else if (expression_analysis_result.hasWindow() || expression_analysis_result.hasAggregation())
+        {
+            if (expression_analysis_result.hasWindow())
+            {
+                const auto & window_analysis_result = expression_analysis_result.getWindow();
+                if (expression_analysis_result.hasAggregation())
+                    addExpressionStep(query_plan, window_analysis_result.before_window_actions, "Before window functions", result_actions_to_execute);
```

Comment:
Haven't we already evaluated it on line 1222?

Reply:
We evaluated it on 1222 only if there were no aggregation. If there is aggregation we must finish aggregation on initiator and then evaluate expressions before window functions and windows.